BaikalDataGenerator/Source/data_generator_impl.cpp

```diff
@@ -0,0 +1,326 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+********************************************************************/
+
+#include "data_generator_impl.h"
+
+#include "filesystem.h"
+#include "devices.h"
+#include "logging.h"
+#include "utils.h"
+
+#include "Baikal/Output/clwoutput.h"
+#include "Baikal/Renderers/monte_carlo_renderer.h"
+#include "Baikal/Renderers/renderer.h"
+#include "Baikal/RenderFactory/clw_render_factory.h"
+
+#include "OpenImageIO/imageio.h"
+
+#include "XML/tinyxml2.h"
+
+#include <algorithm>
+#include <fstream>
+#include <memory>
+#include <vector>
+
+
+struct OutputInfo
+{
+    Baikal::Renderer::OutputType type;
+    std::string name;
+    // channels number can be only 1 or 3
+    int channels_num;
+};
+
+namespace {
+
+// if you need to add new output for saving to disk
+// for iterations number counted in spp.xml file
+// just put its description in this collection
+const std::vector<OutputInfo> kMultipleIteratedOutputs
+{
+    {Baikal::Renderer::OutputType::kColor, "color", 3},
+    {Baikal::Renderer::OutputType::kAlbedo, "albedo", 3},
+    {Baikal::Renderer::OutputType::kGloss, "gloss", 1}
+};
+
+// if you need to add new output for saving to disk
+// only for the one time
+// just put its description in this collection
+const std::vector<OutputInfo> kSingleIteratedOutputs
+{
+    {Baikal::Renderer::OutputType::kViewShadingNormal, "view_shading_normal", 3},
+    {Baikal::Renderer::OutputType::kDepth, "view_shading_depth", 1}
+};
+
+} // namespace
+
+
+DataGeneratorImpl::DataGeneratorImpl(SceneObject* scene,
+                                     size_t output_width,
+                                     size_t output_height,
+                                     std::uint32_t num_bounces,
+                                     unsigned device_idx)
+    : m_scene(scene),
+      m_width(static_cast<std::uint32_t>(output_width)),
+      m_height(static_cast<std::uint32_t>(output_height)),
+      m_num_bounces(num_bounces),
+      m_device_idx(device_idx)
+{
+    assert(m_width);
+    assert(m_height);
+    assert(num_bounces);
```

Comment:
useless asserts, because checks are done in the data_generator.cpp

Reply:
Done